### ISSUE TITLE
Enable TCP & ADQ tests in SPDK CI

### ISSUE THREAD

**Issue by karlatec** [spdk/spdk#1851]
This is an enhancement issue which goal is to enable CI tests for NVMe-oF + ADQ using physical CVL E810 NICs.

What needs to be done in order to make this work:
- [x] Install CVL E810 NICs in the CI environment
- [x] Install ICE driver with ADQ option enabled in CI environment
- [x] Prepare CI job to run SPDK NVMe-oF TCP test suite on top of physical systems instead of virtual machines (e.g. nvmf-phy-autotest)
- [x] Fix any issues that show up when running TCP test suite on physical systems
- [x] Modify test/nvmf TCP scripts to run tests with physical interfaces instead of virtual veth interfaces. Sub-tasks include:
  - [x] Detect any physical NIC interfaces available in test system. Note: this should probably be done only for specified driver, for example ICE.
    Use SPDK_TEST_NVMF_DRIVER test flag to specify the driver for autotest.sh script. (used another name "found_drivers" instead)
  - [x] Create a fallback routine; if no physical NICs for given driver are detected then fall back to nvmf_veth_init() function and do the tests using only veth interfaces.
  - [x] Modify or extend nvmf_veth_init() function:
      - [x] if one capable NIC interface is found then use it as NVMF_TARGET_INTERFACE instead of veth interface. (because associated NVMF_FIRST_TARGET_IP is used most frequently in scripts, so it gives us most coverage)
      - [x] if second capable NIC interface is found then use it as NVMF_TARGET_INTERFACE2 instead of veth interface.
      - [x] if second capable NIC interfaces is found then use it as NVMF_INITIATOR_INTERFACE instead of veth interface. (because associated NVMF_INITIATOR_IP is used lest frequently in scripts)
	  - [x] for any NVMF_*_INTERFACE above, if no corresponding physical NIC interface can be used then use veth instead.
- [x] Once TCP tests are successfully passing using physical NICs after modifying the scripts - create ADQ specific use-cases.

When starting work on some specific task please leave a comment with it's name and a link to Gerrit change (for script changes) so that others can review and track it's progress.
Once it's finished (successfully passed review and CI tests, and is ready to merge) please leave a comment as well, so I can mark the task as "done" in the list above.

As discussed in meeting:

I will take care of following:
- Install CVL E810 NICs in the CI environment
- Install ICE driver with ADQ option enabled in CI environment
- Prepare CI job to run SPDK NVMe-oF TCP test suite on top of physical systems instead of virtual machines (e.g. nvmf-phy-autotest)

@wanqunintel and team will take care of the tasks:
- "Modify test/nvmf TCP scripts to run tests with physical interfaces instead of virtual veth interfaces" (including sub-tasks)
- "Once TCP tests are successfully passing using physical NICs after modifying the scripts - create ADQ specific use-cases."

I am leaving "Fix any issues that show up when running TCP test suite on physical systems" not assigned, because anyone can encounter an issue. If you do then please report it as a new issue ASAP and if it's blocking - link it in this thread.

**Comment by wanqunintel** [comment]
@karol, Hailiang has found that there is no physical machine in the CI par-patch node configured to run with TCP (with E810 physical NIC), Would you provide the node name so that we can submit patch to config it to run with TCP on that machine. I added Hailiang into the thread.

**Comment by karlatec** [comment]
> here is no physical machine in the CI par-patch node configured to run with TCP (with E810 physical NIC)

Correction - there is no CI per-patch job configured to run TCP tests on nodes with E810 NICs. Please let me finish my part of the job (see the tasks listed in the description) and I will comment here when it's ready. 
Just to be clear - lack of CI per-patch job is not a blocker for modifying the test scripts.

**Comment by Karol** [comment]
@wanghailiangx Please use: @karlatec not @Karol
because I get allso notifications

**Comment by karlatec** [comment]
@wanghailiangx  small commit from me which should make detecting desired NIC interfaces easier: https://review.spdk.io/gerrit/c/spdk/spdk/+/7015

Feel free to cherry-pick and/or modify the contents.

**Comment by wanqunintel** [comment]
@karlatec , just would like to know, without that CI machine ready, how can we test the scripts? @wanghailiangx has also posted that there is a autorun_spdk.conf file to config to run with TCP transport.

**Comment by karlatec** [comment]
@wanqunintel AFAIK you have a locally available server with a E810 NIC, so running the tests locally should not be a problem. In fact - if hardware is available locally then all changes should probably be tested as much as possible by the owner of the change, so that they're in good shape when submitted. This helps to reduce CI cycles.

**Comment by karlatec** [comment]
> Install CVL E810 NICs in the CI environment

Done. We have E810's installed in a bunch of CI systems

**Comment by wanqunintel** [comment]
@karlatec , so my understanding is that we let the script running (after address the comments) here locally on our machine nightly. Just correct me if i was wrong. thanks.

**Comment by karlatec** [comment]
@wanqunintel Please use your local system **for development work on the scripts**. Upload the changes frequently to Gerrit for review, so that we can get them in shape quickly. After we finish the reviewing cycle and the patches are merged they will be run as part of usual  CI pipeline.

We will **not** run these tests as separate entity in CI.

**Comment by karlatec** [comment]
> Install ICE driver with ADQ option enabled in CI environment 

This is of course also ready. Forgot to add that yesterday with previous point.

> Prepare CI job to run SPDK NVMe-oF TCP test suite on top of physical systems instead of virtual machines (e.g. nvmf-phy-autotest)

Is blocked. This patch must be merged before I can deploy a CI job : https://review.spdk.io/gerrit/c/spdk/spdk/+/6998

**Comment by jimharris** [comment]
> @wanqunintel Please use your local system **for development work on the scripts**. Upload the changes frequently to Gerrit for review, so that we can get them in shape quickly. After we finish the reviewing cycle and the patches are merged they will be run as part of usual CI pipeline.

I agree.

The typical process is to develop and test the scripts locally, push the scripts to Gerrit for review, and only enable in CI after the patches are merged.

@wanqunintel - can you push your current scripts out to Gerrit to start reviews?

Thanks,

Jim

**Comment by Comphix** [comment]
Hi Jim, just added you on to the patch review. We've already pushed the patch out for review to run the TCP on the physical CVL NIC. If the nic driver is ADQ enabled, it will cover the TCP with ADQ capability. So here, we also listed some pre-work items.

**Comment by Comphix** [comment]
Added the patch from Hailiang https://review.spdk.io/gerrit/c/spdk/spdk/+/6932 here to make the NVMe-oF testing running on physical CVL NIC.

**Comment by wanqunintel** [comment]
@karlatec i saw that you have enabled the CI with ice driver, which driver version is it? And pls. confirm that the ADQ configured correctly.

**Comment by Comphix** [comment]
To make clear

Patch https://review.spdk.io/gerrit/c/spdk/spdk/+/6932 is to run TCP test using the E810 if there. It will not choose which NIC from CVL and others.

"test/nvmf_tcp: allow tcp cases running in E810 NIC instead of virtual eth".

For below items listed in the description:

"Modify test/nvmf TCP scripts to run tests with physical interfaces instead of virtual veth interfaces. Sub-tasks"

@karlatec , Anna and Hailiang will enable the NVMe over TCP on our server for CVL NIC only and patch is already pushed out for review. For the other subtasks listed below, as we talked in last meeting, Anna and Hailing will not work on these, they will only enable the NVMe over TCP testing running on CVL NIC (we will only install CVL NIC on our server). Will you pick up below work items to enable TCP testing in the per-patch CI system? 

a. Detect any physical NIC interfaces available in test system. Note: this should probably be done only for specified driver, for example ICE.
Use SPDK_TEST_NVMF_DRIVER test flag to specify the driver for autotest.sh script.
b. Create a fallback routine; if no physical NICs for given driver are detected then fall back to nvmf_veth_init() function and do the tests using only veth interfaces.
 Modify or extend nvmf_veth_init() function:
 if one capable NIC interface is found then use it as NVMF_TARGET_INTERFACE instead of veth interface. (because associated NVMF_FIRST_TARGET_IP is used most frequently in scripts, so it gives us most coverage)
 if second capable NIC interface is found then use it as NVMF_TARGET_INTERFACE2 instead of veth interface.
 if second capable NIC interfaces is found then use it as NVMF_INITIATOR_INTERFACE instead of veth interface. (because associated NVMF_INITIATOR_IP is used lest frequently in scripts)
 for any NVMF_*_INTERFACE above, if no corresponding physical NIC interface can be used then use veth instead.

**Comment by wanghailiangx** [comment]
Patch https://review.spdk.io/gerrit/c/build_pool/jenkins_build_pool/+/7076 is to get an automated nvmf_tcp test log on physical E810 card.

**Comment by karlatec** [comment]
> Prepare CI job to run SPDK NVMe-oF TCP test suite on top of physical systems instead of virtual machines (e.g. nvmf-phy-autotest)

Done. CI job `nvmf-tcp-phy-autotest` is now run as part of CI per-patch routine.

**Comment by wanghailiangx** [comment]
Updated https://review.spdk.io/gerrit/c/spdk/spdk/+/6932
I have added a function nvmf_phy_init() which uses netns for current TCP testing, and we have tested it locally. Now we can see TCP packets between physical ports. The premise is that we need to connect a cable between the two ports of the E810 network card.

**Comment by wanqunintel** [comment]
@karlatec , with the patch, we have marked the tasks that have been completed. For others that not marked yet, we will continue with the tasks. Thanks.

**Comment by karlatec** [comment]
>  Fix any issues that show up when running TCP test suite on physical systems

Marking this one as done, as the tests have been running in CI for some time and I guess it's stable now.

**Comment by wanghailiangx** [comment]
@Karol Do you want to review this patch? https://review.spdk.io/gerrit/c/spdk/spdk/+/7573.
Customers get spdk code and run nvme_tcp, there may be IP lost problem. They don't know to shut down NetworkManager like CI, so we need to deal with it automatically in the script.

**Comment by karlatec** [comment]
@wanghailiangx  @wanqunintel could we update the list in issue description to match the actual status of the task?

https://review.spdk.io/gerrit/c/spdk/spdk/+/7573 is stuck in review with -1 unresolved comments.

**Comment by wanqunintel** [comment]
@karlatec , thanks for bring this up, i think these are almost done at the very early stage after Hailiang has enabled the NVMe-oF with TCP transport on the CI. @wanghailiangx , would you pls. help to confirm about this. And for this patch, https://review.spdk.io/gerrit/c/spdk/spdk/+/7573, it seems a very old one, do we still need it as i see CI currently didn't have any issue related with IP lost now.

**Comment by wanghailiangx** [comment]
@karlatec  https://review.spdk.io/gerrit/c/spdk/spdk/+/7573 is a very old one, I abandoned it just now.

`if second capable NIC interface is found then use it as NVMF_TARGET_INTERFACE2 instead of veth interface.`
This hasn't come true yet. There's no plan yet.

`for any NVMF_*_INTERFACE above, if no corresponding physical NIC interface can be used then use veth instead.`
This has been implemented, and CI does so.

`Once TCP tests are successfully passing using physical NICs after modifying the scripts - create ADQ specific use-cases.`
This will be implemented by the following script, https://review.spdk.io/gerrit/c/spdk/spdk/+/7476, which is nearing the end.

@wanqunintel I have no access to edit the lines, please help to modify, thanks.

**Comment by wanqunintel** [comment]
Just updated above tasks per Hailiang's comments, @karlatec

**Comment by karlatec** [comment]
Thanks for the update!
btw https://review.spdk.io/gerrit/c/spdk/spdk/+/7476 this one also has a lot of unresolved comments

**Comment by wanqunintel** [comment]
We are keep updating this patch, currently, we are trying to close the gap with adq and baseline yet, will update the patchset after it's stable, stay tuned. :)

**Comment by wanghailiangx** [comment]
@wanqunintel The ADQ patch has been merged, https://review.spdk.io/gerrit/c/spdk/spdk/+/7476. Please update the item:
- [ ] `Once TCP tests are successfully passing using physical NICs after modifying the scripts - create ADQ specific use-cases. `

**Comment by karlatec** [comment]
Points in the list are not updated, but I can see `NVMF_TARGET_INTERFACE2` being set and used in test/nvmf scripts. I think this enhancement is done and was for some time now. ADQ tests are present and functional in our tests and CI. If there are no objections I'd like to close this issue.

Just so that I don't act purely on my own - @mikeBashStuff can you take a look and reply if you agree?

**Comment by mikeBashStuff** [comment]
@karlatec Agreed, this can be closed.
